lib/linter/code-path-analysis/code-path-state.js

```diff
@@ -12,13 +12,594 @@
+     *      conditional expression, this is `"test"`; otherwise, this is `"loop"`.
+     * @param {boolean} isForkingAsResult Indicates if the result of the choice
+     *      creates a fork.
+     * @param {ForkContext} forkContext The containing `ForkContext`.
+     */
+    constructor(upperContext, kind, isForkingAsResult, forkContext) {
+
+        /**
+         * The previous `ChoiceContext`
+         * @type {ChoiceContext}
+         */
+        this.upper = upperContext;
+
+        /**
+         * The kind of choice. If it's a logical or assignment expression, this
+         * is `"&&"` or `"||"` or `"??"`; if it's an `if` statement or
+         * conditional expression, this is `"test"`; otherwise, this is `"loop"`.
+         * @type {string}
+         */
+        this.kind = kind;
+
+        /**
+         * Indicates if the result of the choice forks the code path.
+         * @type {boolean}
+         */
+        this.isForkingAsResult = isForkingAsResult;
+
+        /**
+         * The fork context for the `true` path of the choice.
+         * @type {ForkContext}
+         */
+        this.trueForkContext = ForkContext.newEmpty(forkContext);
+
+        /**
+         * The fork context for the `false` path of the choice.
+         * @type {ForkContext}
+         */
+        this.falseForkContext = ForkContext.newEmpty(forkContext);
+
+        /**
+         * The fork context for the right side of the `??` path of the choice.
+         * @type {ForkContext}
+         */
+        this.qqForkContext = ForkContext.newEmpty(forkContext);
+
+        /**
+         * Indicates if any of `trueForkContext`, `falseForkContext`, or
+         * `qqForkContext` have been updated with segments from a child context.
+         * @type {boolean}
+         */
+        this.processed = false;
+    }
+
+}
+
+/**
+ * Base class for all loop contexts.
+ */
+class LoopContextBase {
+
+    /**
+     * Creates a new instance.
+     * @param {LoopContext} upperContext The previous `LoopContext`.
+     * @param {string} type The AST node's `type` for the loop.
+     * @param {string|null} label The label for the loop from an enclosing `LabeledStatement`.
+     * @param {BreakContext} breakContext The context for breaking the loop.
+     */
+    constructor(upperContext, type, label, breakContext) {
+
+        /**
+         * The previous `LoopContext`.
+         * @type {LoopContext}
+         */
+        this.upper = upperContext;
+
+        /**
+         * The AST node's `type` for the loop.
+         * @type {string}
+         */
+        this.type = type;
+
+        /**
+         * The label for the loop from an enclosing `LabeledStatement`.
+         * @type {string|null}
+         */
+        this.label = label;
+
+        /**
+         * The hardcoded literal boolean test condition for
+         * the loop. Used to catch infinite or skipped loops.
+         * @type {boolean|undefined}
+         */
+        this.test = void 0;
```

Comment:
This isn't a common property for all loop types, for-in and for-of contexts didn't have this property in the previous version.

Reply:
Ooh, nice catch!